### ISSUE TITLE
Publish package to NPM

### ISSUE THREAD

**Issue by alferov** [Narzerus/angular-permission#74]
Hey @Narzerus,

Could you publish this package to NPM? It's crucial for everyone who uses alternative workflows such as browserify and webpack.

Thank you.

**Comment by Narzerus** [comment]
Will do, sorry for the delay on this one, been pretty busy

**Comment by Narzerus** [comment]
@alferov So I started working on this and realized I've never published a front-end package on NPM, can you explain to me what happens here with the dependencies?

This doesn't make a lot of sense from my point of view

**Comment by alferov** [comment]
@Narzerus, just take a look at [browserify](http://browserify.org/) documentation (webpack handles this almost in the same way):
> Browserify lets you require('modules') in the browser by bundling up all of your dependencies.

Once the package is included in package.json as a dependency, you can "require" it just as a regular node.js package in FE JavaScript code. Browserify parses that "require" statements,  looks for according modules in `node_modules`, grabs and injects it (it looks for scripts specified in "main" package.json property) as [source maps](http://benclinkinbeard.com/posts/how-browserify-works/#the-module-map) automatically. 
As for dependencies: it should be required [manually](https://github.com/jakemmarsh/angularjs-gulp-browserify-boilerplate/blob/master/app/js/main.js), so you don't need to include it as dependency in package.json. BTW, [angularjs-gulp-browserify-boilerplate](https://github.com/jakemmarsh/) is a great package to get started with angular and browserify.
Unfortunately, I've never published FE packages on npm, but you can use [Smart-Table package.json](https://github.com/lorenzofox3/Smart-Table/blob/master/package.json) as a starting point.

**Comment by bajramemini** [comment]
Any updates on this?

Kind Regards

**Comment by Narzerus** [comment]
Sorry for the massive delay people, been busy as hell. I'm going to implement this ASAP.

I think I'll have to give merge rights to a couple of developers so that this package can stay with responsive support.

**Comment by Narzerus** [comment]
Ok, so I'm having some kind of problem here in case anyone is willing to help me with this one.

First: I had to violate **semver** and bump the production version to `v1.1.0` (at some point in time I went full-retard some months ago and pushed the packaged to npm with version `1.0.6`.

I'm currently trying to publish to NPM the newest version [this one](https://github.com/Narzerus/angular-permission/releases/tag/v1.1.0)

But I'm getting this error:

```
0 info it worked if it ends with ok
1 verbose cli [ '/Users/rafaelvidaurre/.nvm/versions/node/v0.12.7/bin/node',
1 verbose cli   '/Users/rafaelvidaurre/.nvm/versions/node/v0.12.7/bin/npm',
1 verbose cli   'publish',
1 verbose cli   'angular-permission' ]
2 info using npm@2.11.3
3 info using node@v0.12.7
4 verbose publish [ 'angular-permission' ]
5 silly cache add args [ 'angular-permission', null ]
6 verbose cache add spec angular-permission
7 silly cache add parsed spec { raw: 'angular-permission',
7 silly cache add   scope: null,
7 silly cache add   name: 'angular-permission',
7 silly cache add   rawSpec: '',
7 silly cache add   spec: '*',
7 silly cache add   type: 'range' }
8 silly addNamed angular-permission@*
9 verbose addNamed "*" is a valid semver range for angular-permission
10 silly addNameRange { name: 'angular-permission', range: '*', hasData: false }
11 silly mapToRegistry name angular-permission
12 silly mapToRegistry using default registry
13 silly mapToRegistry registry https://registry.npmjs.org/
14 silly mapToRegistry uri https://registry.npmjs.org/angular-permission
15 verbose addNameRange registry:https://registry.npmjs.org/angular-permission not in flight; fetching
16 verbose request uri https://registry.npmjs.org/angular-permission
17 verbose request no auth needed
18 info attempt registry request try #1 at 4:14:29 PM
19 verbose request id 37e76020e89a63f4
20 http request GET https://registry.npmjs.org/angular-permission
21 http 200 https://registry.npmjs.org/angular-permission
22 silly get cb [ 200,
22 silly get   { server: 'CouchDB/1.5.0 (Erlang OTP/R16B03)',
22 silly get     etag: '"D7TNHCVMCRXZSHPVHTMV4T0Q3"',
22 silly get     'content-type': 'application/json',
22 silly get     'cache-control': 'max-age=60',
22 silly get     'content-length': '8841',
22 silly get     'accept-ranges': 'bytes',
22 silly get     date: 'Wed, 30 Sep 2015 19:14:52 GMT',
22 silly get     via: '1.1 varnish',
22 silly get     age: '0',
22 silly get     connection: 'keep-alive',
22 silly get     'x-served-by': 'cache-sjc3133-SJC',
22 silly get     'x-cache': 'HIT',
22 silly get     'x-cache-hits': '1',
22 silly get     'x-timer': 'S1443640492.270987,VS0,VE86',
22 silly get     vary: 'Accept' } ]
23 verbose get saving angular-permission to /Users/rafaelvidaurre/.npm/registry.npmjs.org/angular-permission/.cache.json
24 silly addNameRange number 2 { name: 'angular-permission', range: '*', hasData: true }
25 silly addNameRange versions [ 'angular-permission', [ '1.0.6' ] ]
26 silly addNamed angular-permission@1.0.6
27 verbose addNamed "1.0.6" is a plain semver version for angular-permission
28 silly cache afterAdd angular-permission@1.0.6
29 verbose afterAdd /Users/rafaelvidaurre/.npm/angular-permission/1.0.6/package/package.json not in flight; writing
30 verbose afterAdd /Users/rafaelvidaurre/.npm/angular-permission/1.0.6/package/package.json written
31 silly publish { dependencies: {},
31 silly publish   devDependencies:
31 silly publish    { grunt: '^0.4.5',
31 silly publish      'grunt-contrib-clean': '^0.5.0',
31 silly publish      'grunt-contrib-concat': '^0.4.0',
31 silly publish      'grunt-contrib-connect': '^0.7.1',
31 silly publish      'grunt-contrib-copy': '^0.5.0',
31 silly publish      'grunt-contrib-jshint': '^0.10.0',
31 silly publish      'grunt-contrib-watch': '^0.6.1',
31 silly publish      'grunt-filerev': '^0.2.1',
31 silly publish      'grunt-karma': '^0.8.3',
31 silly publish      'grunt-newer': '^0.7.0',
31 silly publish      karma: '^0.12.17',
31 silly publish      'karma-jasmine': '^0.1.5',
31 silly publish      'karma-phantomjs-launcher': '^0.1.4',
31 silly publish      'load-grunt-tasks': '^0.4.0',
31 silly publish      'time-grunt': '^0.3.1' },
31 silly publish   scripts: { test: 'grunt karma:unit' },
31 silly publish   name: 'angular-permission',
31 silly publish   description: '![alt tag](https://travis-ci.org/Narzerus/angular-permission.svg?branch=master)',
31 silly publish   version: '1.0.6',
31 silly publish   main: 'dist/angular-permission.js',
31 silly publish   directories: { test: 'test' },
31 silly publish   repository:
31 silly publish    { type: 'git',
31 silly publish      url: 'git+https://github.com/Narzerus/angular-permission.git' },
31 silly publish   keywords: [ 'permission', 'angular', 'routes', 'ui-router', 'access' ],
31 silly publish   author: { name: 'Rafael Vidaurre' },
31 silly publish   license: 'MIT',
31 silly publish   bugs: { url: 'https://github.com/Narzerus/angular-permission/issues' },
31 silly publish   homepage: 'https://github.com/Narzerus/angular-permission#readme',
31 silly publish   readme: '![alt tag](https://travis-ci.org/Narzerus/angular-permission.svg?branch=master)\n\nPermission\n==========\n*Role and permission based authentication on routes as simple as it can get.*\n\n- Requires you to use [ui-router](https://github.com/angular-ui/ui-router) as your router module.\n\nPermission is the gatekeeper for your routes\n--------------------------------------------\nPermission helps you gain control of your routes, by using simple concepts for you to decide who can access them.\nI\'ve seen plenty of big fat tutorials on access control implementations, and they can be quite overwhelming. So I bring you a simple, powerful, straightforward solution.\n\nImportant\n---------\nPlease remember this project is very new, I wouldn\'t recommend yet using this in a big project just yet as, like any new project, might drastically change over time.\n\nInstall via bower\n-----------------\n```\nbower install angular-permission --save\n```\n\nInclude to your dependencies\n----------------------------\n```javascript\nangular.module(\'yourModule\', [..., \'permission\']);\n```\n\nSetting route permissions/roles\n-------------------------------\nThis is how simple Permission makes it for you to define a route which requires authorization.\n\n```javascript\n\n  // We define a route via ui-router\'s $stateProvider\n  $stateProvider\n    .state(\'staffpanel\', {\n      url: \'...\',\n      data: {\n        permissions: {\n          only: [\'admin\', \'moderator\']\n        }\n      }\n    });\n```\n\nYou can either set an `only` or an `except` array.\n\n```javascript\n  // Let\'s prevent anonymous users from looking at a dashboard\n  $stateProvider\n    .state(\'dashboard\', {\n      url: \'...\',\n      data: {\n        permissions: {\n          except: [\'anonymous\']\n        }\n      }\n    });\n```\n\nAnother thing you can do is set a redirect url to which unauthorized sessions will go to.\n\n```javascript\n  $stateProvider\n    .state(\'dashboard\', {\n      url: \'...\',\n      data: {\n        permissions: {\n          except: [\'anonymous\'],\n          redirectTo: \'login\'\n        }\n      }\n    });\n```\n\n\nDefining roles\n--------------------------\nSo, how do yo tell Permission what does \'anonymous\', \'admin\' or \'foo\' mean and how to know if the current user belongs\nto those definitions?\n\nWell, Permission allows you to define different \'roles\' along with the logic that determines if the current\nsession belongs to them.\n\n```javascript\n  // Let\'s imagine we have a User service which has information about the current user in the session\n  // and is undefined if no session is active\n  //\n  // We will define the following roles:\n  // anonymous: When there is not user currenly logged in\n  // normal: A user with isAdmin = false\n  // admin: A user with isAdmin = true\n\n  angular.module(\'fooModule\', [\'permission\', \'user\'])\n    .run(function (Permission, User) {\n      // Define anonymous role\n      Permission.defineRole(\'anonymous\', function (stateParams) {\n        // If the returned value is *truthy* then the user has the role, otherwise they don\'t\n        if (!User) {\n          return true; // Is anonymous\n        }\n        return false;\n      });\n    });\n```\n\nSometimes you will need to call some a back-end api or some other asyncronous task to define the role\nFor that you can use promises\n\n```javascript\n  angular.module(\'barModule\', [\'permission\', \'user\'])\n    .run(function (Permission, User, $q) {\n      Permission\n        // Define user role calling back-end\n        .defineRole(\'user\', function (stateParams) {\n          // This time we will return a promise\n          // If the promise *resolves* then the user has the role, if it *rejects* (you guessed it)\n\n          // Let\'s assume this returns a promise that resolves or rejects if session is active\n          return User.checkSession();\n        })\n        // A different example for admin\n        .defineRole(\'admin\', function (stateParams) {\n          var deferred = $q.defer();\n\n          User.getAccessLevel().then(function (data) {\n            if (data.accessLevel === \'admin\') {\n              deferred.resolve();\n            } else {\n              deferred.reject();\n            }\n          }, function () {\n            // Error with request\n            deferred.reject();\n          });\n\n          return deferred.promise;\n        });\n    });\n```\nAs you can see, Permission is useful wether you want a role-based access control or a permission-based one, as\nit allows you to define this behaviour however you want to.\n\nEvents\n------\n- **$stateChangePermissionStart**: This event is broadcasted before perform authorize.\n\n- **$stateChangePermissionAccepted**: This event is broadcasted when one of the permissions has been accepted and the state changes successfully.\n\n- **$stateChangePermissionDenied**: This event is broadcasted when the access to the target state is not granted (no permissions found on the `only` array or at least one permission found on the `except` array). This is when the state stays the same or is changed based on the `redirectTo` option.\n\n\nCaveats\n=======\nBecause of a bug in ui-router, when using `$urlStateProvider.otherwise` we get an **infinite digest** loop error.\nA workaround was found by [@shaoibmerchant](https://github.com/shoaibmerchant) and it goes like this:\n\n```javascript\n// Normal usage (creates INFDG error)\n$urlRouterProvider.otherwise(\'/somestate\');\n\n// Instead\n$urlRouterProvider.otherwise( function($injector) {\n  var $state = $injector.get("$state");\n  $state.go(\'/somestate\');\n});\n```\n\nTODOS:\n-----\nHelp fill this list with your feature requests\n- More powerful redirection to allow passing state parameters and other useful stuff ui-router provides. Ideas anyone?\n- Inheritance (example: \'admin\' inherits from \'user\')\n- Role validation caching?\n\nContributing\n============\nThis project is still in diapers and I would love your feedback / help in making this a great module\nfor angular developers to use.\n\nThe correct way to contribute is:\n  1. Create a branch from the `development` branch\n  2. Implement your new feature\n  3. Submit a pull request to be merged in the `development` branch\n  4. Remember to run `grunt build` before your last commit\n\nAuthor\n======\n- Rafael Vidaurre\n- @narzerus\n- I\'m a full-stack developer currenly working as CTO and Co-Founder at [Finciero](http://www.finciero.com)\n',
31 silly publish   readmeFilename: 'README.md',
31 silly publish   gitHead: 'f86a016c72ce18780b9036fe7c728aefe23c16dd',
31 silly publish   _id: 'angular-permission@1.0.6',
31 silly publish   _shasum: 'e77989164425a3ac011c21dd761272798a5939fa',
31 silly publish   _from: 'angular-permission@*' }
32 info prepublish angular-permission@1.0.6
33 verbose getPublishConfig undefined
34 silly mapToRegistry name angular-permission
35 silly mapToRegistry using default registry
36 silly mapToRegistry registry https://registry.npmjs.org/
37 silly mapToRegistry uri https://registry.npmjs.org/angular-permission
38 verbose publish registryBase https://registry.npmjs.org/
39 silly publish uploading /Users/rafaelvidaurre/.npm/angular-permission/1.0.6/package.tgz
40 verbose request uri https://registry.npmjs.org/angular-permission
41 verbose request sending authorization for write operation
42 info attempt registry request try #1 at 4:14:52 PM
43 http request PUT https://registry.npmjs.org/angular-permission
44 http 403 https://registry.npmjs.org/angular-permission
45 verbose headers { 'content-type': 'application/json',
45 verbose headers   'cache-control': 'max-age=60',
45 verbose headers   'content-length': '95',
45 verbose headers   'accept-ranges': 'bytes',
45 verbose headers   date: 'Wed, 30 Sep 2015 19:14:53 GMT',
45 verbose headers   via: '1.1 varnish',
45 verbose headers   connection: 'keep-alive',
45 verbose headers   'x-served-by': 'cache-sjc3128-SJC',
45 verbose headers   'x-cache': 'MISS',
45 verbose headers   'x-cache-hits': '0',
45 verbose headers   'x-timer': 'S1443640493.080011,VS0,VE456' }
46 verbose request invalidating /Users/rafaelvidaurre/.npm/registry.npmjs.org/angular-permission on PUT
47 error publish Failed PUT 403
48 verbose stack Error: "You cannot publish over the previously published version 1.0.6." : angular-permission
48 verbose stack     at CachingRegistryClient.<anonymous> (/Users/rafaelvidaurre/.nvm/versions/node/v0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:247:14)
48 verbose stack     at Request._callback (/Users/rafaelvidaurre/.nvm/versions/node/v0.12.7/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:170:14)
48 verbose stack     at Request.self.callback (/Users/rafaelvidaurre/.nvm/versions/node/v0.12.7/lib/node_modules/npm/node_modules/request/request.js:354:22)
48 verbose stack     at Request.emit (events.js:110:17)
48 verbose stack     at Request.<anonymous> (/Users/rafaelvidaurre/.nvm/versions/node/v0.12.7/lib/node_modules/npm/node_modules/request/request.js:1207:14)
48 verbose stack     at Request.emit (events.js:129:20)
48 verbose stack     at IncomingMessage.<anonymous> (/Users/rafaelvidaurre/.nvm/versions/node/v0.12.7/lib/node_modules/npm/node_modules/request/request.js:1153:12)
48 verbose stack     at IncomingMessage.emit (events.js:129:20)
48 verbose stack     at _stream_readable.js:908:16
48 verbose stack     at process._tickCallback (node.js:355:11)
49 verbose statusCode 403
50 verbose pkgid angular-permission
51 verbose cwd /Users/rafaelvidaurre/Code/angular-permission
52 error Darwin 14.3.0
53 error argv "/Users/rafaelvidaurre/.nvm/versions/node/v0.12.7/bin/node" "/Users/rafaelvidaurre/.nvm/versions/node/v0.12.7/bin/npm" "publish" "angular-permission"
54 error node v0.12.7
55 error npm  v2.11.3
56 error code E403
57 error "You cannot publish over the previously published version 1.0.6." : angular-permission
58 error If you need help, you may report this error at:
58 error     <https://github.com/npm/npm/issues>
59 verbose exit [ 1, true ]

```

**Comment by Narzerus** [comment]
Package's version 1.1.1 is now available at NPM
